Parity Symmetry

mtsch opened the pull request.

mtsch committed:
Add `ParitySymmetry`

mtsch committed:
Add docstring

mtsch committed:
Update docstring

mtsch committed:
Add more tests

joachimbrand commented on src/Hamiltonians/ParitySymmetry.jl:
```diff
@@ -0,0 +1,126 @@
+    weight = left_weight(address)
+    direction = weight ≥ 0 ? -1 : 1
+    if !even && weight == 0
+        error("Even starting address can't be used with odd `ParitySymmetry`")
```
```suggestion
        throw(ArgumentError("Even starting address can't be used with odd `ParitySymmetry`"))
```

joachimbrand commented on src/Hamiltonians/ParitySymmetry.jl:
```diff
@@ -0,0 +1,126 @@
+    address = starting_address(hamiltonian)
+
+    if !isodd(num_modes(address))
+        error("Starting address must have an odd number of modes")
```
```suggestion
        throw(ArgumentError("Starting address must have an odd number of modes"))
```
See, e.g. https://discourse.julialang.org/t/error-vs-throw-argumenterror/18915

joachimbrand replied:
Actually, what is the problem with even number of modes. The midpoint would be a non-integer, but is there a real problem? Parity should still exist and make sense.

mtsch replied:
Ok, I'll use `ArgumentError`s from now on. 

The problem with the number of modes is in momentum space, where reversing an address with an even number of modes gives you an address with a different momentum. The correct way to do it there would be to leave the last mode alone, but that would break real space. We could get around this if we could distinguish momentum space addresses from real space ones.

mtsch committed:
Apply suggestions from code review

Co-authored-by: Joachim Brand <j.brand@massey.ac.nz>

mtsch committed:
Replace some more `error`s

mtsch committed:
Change failing doctest

mtsch committed:
Fix failing doctest again (?)

joachimbrand commented:
I'm confused by this:
```
julia> Matrix(ParitySymmetry(ham, odd=true))|>eigvals
ERROR: ArgumentError: Even starting address can't be used with odd `ParitySymmetry`
Stacktrace:
 [1] #ParitySymmetry#65
   @ ~/git/Rimu/src/Hamiltonians/ParitySymmetry.jl:75 [inlined]
 [2] top-level scope
   @ none:1

julia> starting_address(ham)
BoseFS{4,5}((0, 2, 1, 0, 1))

julia> import Rimu.Hamiltonians: left_weight

julia> starting_address(ham)|>left_weight
0
```
The starting address is clearly not even, and an odd-parity state
```
BoseFS{4,5}((0, 2, 1, 0, 1)) - BoseFS{4,5}((1, 0, 1, 2, 0))
```
can be easily constructed and seems legit.

So it looks like `left_weight` returns zero on some non-even states, which are then mistakenly classified as even. This seems wrong. What I don't understand is why this doesn't seem to be causing serious damage.

mtsch committed:
Simplify finding left-heavy addresses

mtsch commented:
> I'm confused by this:
> 
> ```
> julia> Matrix(ParitySymmetry(ham, odd=true))|>eigvals
> ERROR: ArgumentError: Even starting address can't be used with odd `ParitySymmetry`
> Stacktrace:
>  [1] #ParitySymmetry#65
>    @ ~/git/Rimu/src/Hamiltonians/ParitySymmetry.jl:75 [inlined]
>  [2] top-level scope
>    @ none:1
> 
> julia> starting_address(ham)
> BoseFS{4,5}((0, 2, 1, 0, 1))
> 
> julia> import Rimu.Hamiltonians: left_weight
> 
> julia> starting_address(ham)|>left_weight
> 0
> ```
> 
> The starting address is clearly not even, and an odd-parity state
> 
> ```
> BoseFS{4,5}((0, 2, 1, 0, 1)) - BoseFS{4,5}((1, 0, 1, 2, 0))
> ```
> 
> can be easily constructed and seems legit.
> 
> So it looks like `left_weight` returns zero on some non-even states, which are then mistakenly classified as even. This seems wrong. What I don't understand is why this doesn't seem to be causing serious damage.

Fixed and added this case as a test.

mtsch committed:
Fix doctest

mtsch committed:
Remove `left_heavy`

joachimbrand committed:
docstring for ParitySymmetry

joachimbrand reviewed (dismissed):
What an awesome new feature!

I'm happy now with the code. Please have a look at the changes I made to the docstring of `ParitySymmetry`!

joachimbrand committed:
changed docstring again
